File: backend/doctors/models/doctor.py
from django.db import models
from model_utils.fields import StatusField, MonitorField
from model_utils import Choices
from phonenumber_field.modelfields import PhoneNumberField

from .abstractCacheRelatedClass import AbstractCacheRelatedClass
from .doctorCategory import DoctorCategory
from .doctorSpeciality import DoctorSpeciality
from .review import DoctorReview


class Doctor(AbstractCacheRelatedClass):
    """
    A doctor info.
    """

    GENDER_CHOICES = (
        ("M", "Male"),
        ("F", "Female"),
    )

    STATUS = Choices("PENDING_APPROVAL", "APPROVED", "REJECTED", "DELETED", "RETIRED", "PASSED AWAY")

    full_name = models.CharField(max_length=100)
    gender = models.CharField(
        max_length=1, choices=GENDER_CHOICES, default=GENDER_CHOICES[0][0]
    )
    category = models.ForeignKey(
        DoctorCategory, blank=True, null=True, on_delete=models.SET_NULL
    )
    specialities = models.ManyToManyField(DoctorSpeciality, blank=True)
    i_care_better = models.URLField(blank=True)
    nancys_nook = models.BooleanField(default=False)
    image = models.ImageField(blank=True, null=True, upload_to="images")
    status = StatusField()
    added_by = models.ForeignKey(
        "users.User", blank=True, null=True, on_delete=models.SET_NULL
    )
    added_at = models.DateTimeField(auto_now_add=True, null=True)
    approved_at = MonitorField(monitor="status", when=["APPROVED"], null=True, blank=True, default=None)  # type: ignore
    rejected_at = MonitorField(monitor="status", when=["REJECTED"], null=True, blank=True, default=None)  # type: ignore
    updated_at = models.DateTimeField(auto_now=True, null=True)
    internal_notes = models.CharField(max_length=100, blank=True, null=True)


    def __str__(self) -> str:
        return str(self.full_name)

    @property
    def reviews(self) -> list[DoctorReview]:
        return list(DoctorReview.objects.filter(doctor=self, status="APPROVED"))

    @property
    def num_of_reviews(self) -> int:
        return len(self.reviews)

    @property
    def avg_rating(self) -> float | None:
        ratings = [review.rating for review in self.reviews if review.rating]
        return (float(sum(ratings)) / len(ratings)) if ratings else None
    

class DoctorLocation(models.Model):
    """
    A doctor location.
    Doctor might have multiple locations.
    """

    doctor = models.ForeignKey(Doctor, on_delete=models.CASCADE)
    hospital_name = models.CharField(
        max_length=200, blank=True, null=True
    )
    long_address = models.CharField(max_length=200, blank=True)
    short_address = models.CharField(max_length=200, blank=True)
    lat = models.DecimalField(max_digits=22, decimal_places=16, blank=True, null=True)
    lng = models.DecimalField(max_digits=22, decimal_places=16, blank=True, null=True)
    phone = PhoneNumberField(null=True, blank=True)
    email = models.EmailField(null=True, blank=True)
    website = models.URLField(null=True, blank=True)
    private_only = models.BooleanField(default=False)

    def __str__(self) -> str:
        return f"{self.long_address} / {self.hospital_name}"

File: frontend/src/types/doctors/doctor.ts
import { userSavedProvidersUrl } from "../../AppRouter";
import { UserInfo } from "../../auth/userInfo";
import { DistanceUnit } from "../../components/utils/DistanceUnit";
import { useReviews } from "../../hooks/doctors/useReviews";
import useGoogleMaps, { Location } from "../../utils/googleMaps/useGoogleMaps";
import { ImageFileOrUrl } from "../Image";
import { DateTime } from "../utils/dateTime";
import { Email } from "../utils/email";
import { ID } from "../utils/id";
import { Phone } from "../utils/phone";
import { Url } from "../utils/url";
import { DoctorReview } from "./review";

export type AbstractLocation = {
    lat?: number;
    lng?: number;
};

export type DoctorLocation = {
    id?: ID;
    hospitalName?: string;
    longAddress?: string;
    shortAddress?: string;
    lat?: number;
    lng?: number;
    phone?: Phone;
    email?: Email;
    website?: Url;
    privateOnly: boolean;
};

export const locationToStr = (location: AbstractLocation) => `location-${Number(location.lat).toFixed(7)}/${Number(location.lng).toFixed(7)}`;

export const sameLocation = (a: AbstractLocation, b: AbstractLocation) => locationToStr(a) === locationToStr(b);

export const newDoctorLocation = (): DoctorLocation => {
    return { privateOnly: false };
};

export const doctorStatuses = ["PENDING_APPROVAL", "APPROVED", "REJECTED", "RETIRED", "PASSED AWAY"] as const;
export type DoctorStatus = (typeof doctorStatuses)[number];
export const doctorStatusToString = (status: DoctorStatus) => status.replaceAll("_", " ");

export const doctorGenders = ["M", "F"] as const;
export type DoctorGender = (typeof doctorGenders)[number];
export const doctorGenderToString = (gender: DoctorGender) =>
    new Map([
        ["M", "Male"],
        ["F", "Female"],
    ]).get(gender) as string;

export type Doctor = {
    id?: ID;
    fullName: string;
    gender: DoctorGender;
    locations: DoctorLocation[];
    category?: string;
    specialities: string[];
    iCareBetter?: Url;
    nancysNook?: boolean;
    image?: ImageFileOrUrl;
    status?: DoctorStatus;
    addedBy?: UserInfo;
    addedAt?: DateTime;
    approvedAt?: DateTime;
    rejectedAt?: DateTime;
    updatedAt?: DateTime;
    numOfReviews?: number;
    avgRating?: number;
};

export const newDoctor = (): Doctor => {
    return {
        fullName: "",
        gender: "M",
        locations: [],
        specialities: [],
    };
};

export function getDoctorNameWithoutPrefix(doctor: Doctor): string {
    const prefixes = ["prof", "dr"];
    let name = doctor.fullName.toLowerCase();
    if (prefixes.some((prefix) => name.startsWith(prefix))) {
        name = name.slice(name.indexOf(" "));
    }
    return name;
}

export function getDoctorLocationDistance(
    doctorLocation: DoctorLocation,
    location: Location,
    distanceUnit?: DistanceUnit
) {
    const { getDistance } = useGoogleMaps();

    return getDistance(location, { lat: Number(doctorLocation.lat!), lng: Number(doctorLocation.lng!) }, distanceUnit);
}

export function getDoctorNearestLocation(doctor: Doctor, location: Location): DoctorLocation | null {
    const doctorLocations = doctor.locations
        .filter((doctorLocation) => doctorLocation.lat && doctorLocation.lng)
        .filter((doctorLocation) => getDoctorLocationDistance(doctorLocation, location) !== undefined)
        .sort((a, b) => getDoctorLocationDistance(a, location) - getDoctorLocationDistance(b, location));

    return doctorLocations.length === 0 ? null : doctorLocations[0];
}

export function getDoctorMinimalDistance(doctor: Doctor, location: Location, distanceUnit: DistanceUnit): number {
    const doctorLocation = getDoctorNearestLocation(doctor, location);

    return doctorLocation === null ? Infinity : getDoctorLocationDistance(doctorLocation, location, distanceUnit);
}

export function getDoctorReviews(doctor: Doctor) {
    const { data } = useReviews();
    return data.filter((review: DoctorReview) => review.doctor.id === doctor.id);
}

export function getDoctorUrl(doctor: Doctor, onlyMyList = false) {
    return `${onlyMyList ? userSavedProvidersUrl : ""}/${doctor.id}/${doctor.fullName.replaceAll(" ", "-")}}`;
}
